### ISSUE TITLE
Drop generic type constraint Query<T> to be T : class, and allow value types

### ISSUE THREAD

**Issue by dzmitry-lahoda** [elastic/elasticsearch-net#2522]
As of now https://github.com/elastic/elasticsearch-net/blob/master/src/Nest/QueryDsl/Query.cs is next
```
public static class Query<T> where T : class
```

Could you drop `class`? I.e.:
```
public static class Query<T> 
```

 So I could write for `stuct`s (i.e. value types)? Any reasons it is impossible?

Use case:
During usage of ElasticSearch as NoSQL storage for audit we do client side aggregation as elastic does not support summaries (sum,count) during usage of `top_hits` aggregation. 
Given different tactics, one of these like take first of all and take last for that day, leads to store millions entities in memory so it could be good to have these to be `struct`s.
Another issues that trying to build LINQ provider for our case is impossible as LINQ provider methods has no `class` constraint.

If `class` consideration was for inheritance and proxy generation in future, do not do it.

Follow VITA ORM or other delegation pattern coding/code generation - use interfaces.

Why not build LINQ to low level client? I like typing and query validation done by NEST, otherwise why I use NEST at all for our usage case.

**Comment by russcam** [comment]
Thanks for opening this @asd-and-Rizzo. I'll need to look into whether this is possible, based on (off the top of my head):

1. the propagation of the generic parameter constraint from other client components i.e. can the constraint be dropped for all
2. breaking binary backwards compatibility

**Comment by asd-and-Rizzo** [comment]
It would not break binary backward compatibility, I think. I did no tests, but if it would break than any drop of constraint in library code would break users - this will contradict all other good rules of CLR about backward compatibility. I think user will work fine with their constraint remained even if new assembly is used with no constraint.

**Comment by asd-and-Rizzo** [comment]
I have done experiments here https://gitlab.com/dzmitry-lahoda/clr/blob/master/enterprise/compatiblity/generic_class_constraint_drop_user/Program.cs

It is safe to drop constraint for both callers of code and for thus who inherited.

**Comment by asd-and-Rizzo** [comment]
I have made first try to drop constraint in all places and failed. I think try to go from most used/root classes and go to childs/users. It is possible to drop constraint or root/most used, while childs/users still have constraint. Go from bottom to top. What are these most important classes?

**Comment by asd-and-Rizzo** [comment]
I am unable to do GroupBy -> Aggregations, https://msdn.microsoft.com/en-us/library/bb534492.aspx , as IGrouping is not class.

**Comment by asd-and-Rizzo** [comment]
As for https://github.com/elastic/elasticsearch-net/issues/1197#issuecomment-69325703 . As soon as ES is used as NoSQL database, LINQ makes perfect fit.

**Comment by KodrAus** [comment]
It seems like there's a few possibly orthogonal things going on here:

- `top_hits` doesn't give you some metrics you'd like, so you're working around it by loading lots of docs and doing some work outside of ES, right? That leads to GC pressure because you've got large bursts of short-lived objects.
- You'd like to be able to use a LINQ API over ES instead of the Query DSL in NEST

Does that sound right @asd-and-Rizzo? 

IMHO removing the `T: class` constraint is a worthwhile change, unless there are plans down the track to do more with it, @russcam or @Mpdreamz would know. That would have to start in `Elasticsearch.Net` rather than `NEST` though. Also I don't really see that as a solution to your problem. Rather than fetching your index in memory on the client and doing your own aggregations could you not use the `sum` or `value_count` aggregations in ES?

**Comment by asd-and-Rizzo** [comment]
@KodrAus, 
As for pure ES side:
1. ElasticSearch in NoSQL data storage.
2. ElasticSearch may be used to store audit data.
3. With audit data we cannot use aggregations as some items will double and triple count. top_hits do not combine with aggregations/sum/counts. 
4. Approach is client side aggregations.
5. ES developers explicitly state that top_hits and aggregations do not work, and suggest to use client side joins.

For C# NEST side of things:
1. I did client side aggregation and join using NEST typed API. 200 lines of code for single query, hard to read for new developers.
2. Layered custom LINQ over NEST. 50 lines of  code. C# devs easy to read already known Select/Where/OrderBy/GroupBy. 
3. Wait a second? .NET Queryable is T. But I have MyQueryable where T: class. So these are different things.
4. Drop class constraint in NEST and get API convergence. Allow to use NEST for more custom query providers.


We do aggregate 3 000 000 objects at once in memory. These are tiny object specifically structured (enums of byte, etc) to fit into memory. Could I use these strait out of NEST if class constrain dropped. Why at all class constrain is presented? I dropped class from promise in NEST code and things compiled and worked. 

Some links for inspirations:
https://siren.solutions/searchplugins/join/
https://github.com/elastic/elasticsearch/issues/16537

**Comment by KodrAus** [comment]
Yeh, I think the advice to aggregate on the client starts to break down a bit when you've got 3,000,000 objects. Even if your document is represented as a `struct` AFAIK there are still a few allocations per hit which are themselves wrapped in a class each, so while you'd be reducing pressure a bit it might not make a significant difference. Is this something you're doing now and are seeing bad GC performance or `OutOfMemory` exceptions?

As for `IQueryable`, I guess that comes down to personal taste. I personally don't buy into the promise of a single query API for all kinds of data sources, but that doesn't mean it isn't useful. What was the issue you had with `ElasticLINQ`? Are you suggesting `NEST` should provide a LINQ provider here?

**Comment by asd-and-Rizzo** [comment]
@KodrAus, 

Why not ElasticLINQ is here https://github.com/ElasticLINQ/ElasticLINQ/issues/103  . I do not need ready to use IQueryable, I need API to build custom IQueryable. Dropping class constraint in NEST is way to give API to build custom LINQ on top. I do not ask NEST to be LINQ provider, I ask NEST do not prevent me from building one.

In our case we have billions of objects from hundreds of indexes are client side joined/aggregated withing 2 hours of data dump from ES. 3M is max size of single collection, we have in memory at once several collection. Hence many GB of heap.

**Comment by Mpdreamz** [comment]
The main reason why `T` is constrained to `class` is for all the inference we do based on types and down the line in expressions inside e.g `Field()`. It also signals that you can't do `Search<string>` since `_source` is always an object. 

C# needs a `struct` constraint badly :(

I'm not opposed to dropping the constraint but would like to explore the ramifications in terms of API misuse. 

I agree with @KodrAus that `struct` on its own won't have that much effect on GC (and his points on `IQueryable`) but NEST should not be in the way of others attempting this.

**Comment by asd-and-Rizzo** [comment]
> The main reason why T is constrained to class is for all the inference we do based on types and down the line in expressions inside e.g Field(). 

I am not aware any of C# rules which differ inference class vs struct. What are these? The only difference in type system I am aware is null vs not null, which is mitigated in many cases by default(T).

>It also signals that you can't do Search<string> since _source is always an object.

I do not get statement about signals. String is class, not struct.

**Comment by codebrain** [comment]
Discussed and will consider making the change to drop the "where T : class" for  user-defined document types.

**Comment by stevejgordon** [comment]
This change has been implemented for `8.0.0-alpha.1`.
